File: 0172-factorial-trailing-zeroes/0172-factorial-trailing-zeroes.java
class Solution {
    /**
     * Calculate the number of trailing zeroes in n! (n factorial).
     * 
     * Trailing zeroes are produced by factors of 10, and 10 = 2 * 5.
     * Since there are always more factors of 2 than 5 in n!,
     * we only need to count the number of factors of 5.
     * 
     * @param n The input number to calculate factorial trailing zeroes
     * @return The number of trailing zeroes in n!
     */
    public int trailingZeroes(int n) {
        int trailingZeroCount = 0;
      
        // Count all factors of 5 in numbers from 1 to n
        // n/5 gives count of numbers divisible by 5
        // n/25 gives count of numbers divisible by 25 (contributing an extra 5)
        // n/125 gives count of numbers divisible by 125 (contributing another extra 5)
        // And so on...
        while (n > 0) {
            n /= 5;  // Integer division to get the count at current power of 5
            trailingZeroCount += n;  // Add the count to total
        }
      
        return trailingZeroCount;
    }
}

File: leetcode/0172-factorial-trailing-zeroes.ts
/**
 * Calculates the number of trailing zeroes in n! (n factorial)
 * 
 * The number of trailing zeroes is determined by the number of times 10 is a factor in n!
 * Since 10 = 2 × 5, and there are always more factors of 2 than 5 in n!,
 * we only need to count the number of factors of 5
 * 
 * @param n - The input number to calculate factorial trailing zeroes for
 * @returns The count of trailing zeroes in n!
 */
function trailingZeroes(n: number): number {
    let trailingZeroCount: number = 0;
  
    // Count all factors of 5 in numbers from 1 to n
    // We divide by 5, 25, 125, etc. to count multiples of 5^1, 5^2, 5^3, etc.
    while (n > 0) {
        // Integer division by 5 to count multiples of current power of 5
        n = Math.floor(n / 5);
      
        // Add the count of multiples to our total
        trailingZeroCount += n;
    }
  
    return trailingZeroCount;
}
